lib/upipe-dvb/upipe_dvbsrc.c

```diff
@@ -0,0 +1,765 @@
+
+UPIPE_HELPER_UPUMP_MGR(upipe_dvbsrc, upump_mgr)
+UPIPE_HELPER_UPUMP(upipe_dvbsrc, upump, upump_mgr)
+
+static void upipe_dvbsrc_log(void *priv, int level, const char *fmt,...)
+{
+    int loglevel;
+    switch (level) {
+    case LOG_WARNING:
+        loglevel = UPROBE_LOG_WARNING;
+        break;
+    case LOG_NOTICE:
+        loglevel = UPROBE_LOG_NOTICE;
+        break;
+    case LOG_DEBUG:
+    case LOG_INFO:
+        loglevel = UPROBE_LOG_DEBUG;
+        break;
+    case LOG_ERR:
+    default:
+        loglevel = UPROBE_LOG_ERROR;
+        break;
+    }
+
+    struct uprobe *uprobe = priv;
+
+    va_list args;
+    va_start(args, fmt);
+
+    size_t len = vsnprintf(NULL, 0, fmt, args);
+    va_end(args);
+    if (len > 0) {
+        char str[len + 1];
+        va_start(args, fmt);
+        vsnprintf(str, len + 1, fmt, args);
+        va_end(args);
+        char *end = str + len - 1;
+        if (isspace(*end)) {
+            *end = '\0';
+        }
+        uprobe_log(uprobe, NULL, loglevel, str);
+    }
+}
+
+/** @internal @This allocates a DVB receiver source pipe.
+ *
+ * @param mgr common management structure
+ * @param uprobe structure used to raise events
+ * @param signature signature of the pipe allocator
+ * @param args optional arguments
+ * @return pointer to upipe or NULL in case of allocation error
+ */
+static struct upipe *upipe_dvbsrc_alloc(struct upipe_mgr *mgr,
+                                        struct uprobe *uprobe,
+                                        uint32_t signature, va_list args)
+{
+    struct upipe *upipe = upipe_dvbsrc_alloc_void(mgr, uprobe, signature, args);
+    struct upipe_dvbsrc *upipe_dvbsrc = upipe_dvbsrc_from_upipe(upipe);
+
+    upipe_dvbsrc->dvb = dvb_dev_alloc();
+    if (!upipe_dvbsrc->dvb) {
+        upipe_err(upipe, "Can't allocate DVB device");
+        upipe_dvbsrc_free_void(upipe);
+        return NULL;
+    }
+
+    if (dvb_dev_find(upipe_dvbsrc->dvb, NULL, NULL) < 0) {
+        upipe_err(upipe, "No DVB devices found");
+        dvb_dev_free(upipe_dvbsrc->dvb);
+        upipe_dvbsrc_free_void(upipe);
+        return NULL;
+    }
+
+    upipe_dvbsrc_init_urefcount(upipe);
+    upipe_dvbsrc_init_uref_mgr(upipe);
+    upipe_dvbsrc_init_ubuf_mgr(upipe);
+    upipe_dvbsrc_init_output(upipe);
+    upipe_dvbsrc_init_upump_mgr(upipe);
+    upipe_dvbsrc_init_upump(upipe);
+    upipe_dvbsrc_init_uclock(upipe);
+    upipe_dvbsrc->uri = NULL;
+    upipe_dvbsrc->demux = NULL;
+    upipe_dvbsrc->frontend = NULL;
+    upipe_dvbsrc->uref = NULL;
+    upipe_dvbsrc->latency = 0;
+    upipe_dvbsrc->current_uref = NULL;
+    upipe_dvbsrc->buffer_size = 96 * MTU; /* Minimum DMA transfer for ddbridge */
+
+    dvb_dev_set_logpriv(upipe_dvbsrc->dvb, 2, upipe_dvbsrc_log, uprobe);
+
+    upipe_throw_ready(upipe);
+    return upipe;
+}
+
+/** @internal */
+static void upipe_dvbsrc_date_and_output(struct upipe *upipe, struct uref *uref, uint64_t systime)
```

Comment:
I think you're doing it the wrong way around. You interpolate the time of the received packets according to the *future* bitrate (up to the next DMA transfer), while you should be interpolating it from the time of the previous DMA transfers, as the packets you get have been received since then.

Reply:
I think you are correct but your approach has the drawback of losing the first set of TS packets.